File: support/src/main/java/com/wuguangxin/base/Task.java
package com.wuguangxin.base;

import com.wuguangxin.utils.Logger;

import java.util.LinkedList;

public abstract class Task<T> {

    LinkedList<T> taskList = new LinkedList<>();

    /**
     * 入栈
     * @param task 入栈的 Activity/Fragment
     */
    abstract void inTask(T task);

    /**
     * 出栈
     * @param task 出栈的 Activity/Fragment
     */
    abstract void outTask(T task);

    /**
     * 清栈，销毁所有 Activity/Fragment
     */
    abstract void clearTask();

    /**
     * 获取栈顶 Activity/Fragment
     * @return 栈顶Activity/Fragment
     */
    abstract T getTopTask();

//	/**
//	 * 回退栈(如从A-B-C-D, 在D界面直接返回到A,现在需要把D,C,B给Finish掉,这时的回退次数num为3)
//	 * @param num 回退的位数
//	 */
//	void backTask(int num);

//	/**
//	 * 回退栈(如从A-B-C-D, 在D界面直接返回到A,会将D,C,B给直接销毁掉)
//	 * @param taskClass 回退到该 Activity/Fragment.class
//	 */
//	void backTask(Class taskClass);

    /**
     * 获取当前栈大小
     * @return
     */
    abstract int getTaskSize();

    /**
     * @return 列表
     */
    abstract LinkedList<T> getTaskList();

    interface ACTION {
        String UP = "UP"; // 上浮
        String IN = "IN"; // 入栈
        String OUT = "OUT"; // 出栈
        String CLEAN = "CLEAN"; // 清栈
    }

    void printTaskInfo(String tag, String className, String msg) {
        String taskTop = "null";
        int taskSize = 0;
        if (taskList != null) {
            taskSize = taskList.size();
            if (taskSize > 0) {
                taskTop = taskList.getFirst().getClass().getSimpleName();
            }
        }
        Logger.d(tag, String.format("%s:%s   TOP:%s   TASK SIZE:%s", msg, className, taskTop, taskSize));
    }

}

File: support/src/main/java/com/wuguangxin/base/Task.kt
package com.wuguangxin.base

import com.wuguangxin.utils.Logger
import java.util.*

abstract class Task<T> {

    val taskList = LinkedList<T>()

    /**
     * 获取栈顶Activity
     * @return 栈顶Activity
     */
    val topTask: T? get() = if (taskSize == 0) null else taskList.first

    /**
     * 获取当前栈大小
     * @return
     */
    val taskSize: Int get() = taskList.size

    /**
     * 入栈
     * @param task 入栈的 Activity/Fragment
     */
    abstract fun inTask(task: T)

    /**
     * 出栈
     * @param task 出栈的 Activity/Fragment
     */
    abstract fun outTask(task: T)

    /**
     * 清栈，销毁所有 Activity/Fragment
     */
    abstract fun clearTask()

    /**
     * 回退栈(如从A-B-C-D, 在D界面直接返回到A,现在需要把D,C,B给Finish掉,这时的回退次数num为3)
     * @param count 回退几次
     */
    //abstract fun backTask(count: Int)

    /**
     * 回退栈(如从A-B-C-D, 在D界面直接返回到A,会将D,C,B给直接销毁掉)
     * @param taskClass 回退到该 Activity/Fragment.class
     */
    //abstract fun backTask(taskClass: Class<T>);

    fun printTaskInfo(tag: String?, className: String?, msg: String?) {
        var taskTop = "null"
        if (taskList.size > 0) {
            taskList.first?.let {
                taskTop = it::class.java.simpleName
            }
        }
        Logger.d(tag, "$msg:$className   TOP:$taskTop   TASK SIZE:$taskSize")
    }

    companion object {
        const val UP = "UP" // 上浮
        const val IN = "IN" // 入栈
        const val OUT = "OUT" // 出栈
        const val CLEAN = "CLEAN" // 清栈
    }
}